source/common/grpc/async_client_impl.cc

```diff
@@ -88,10 +92,17 @@ void AsyncStreamImpl::initialize(bool buffer_body_for_retry) {
       parent_.host_name_.empty() ? parent_.remote_cluster_name_ : parent_.host_name_,
       service_full_name_, method_name_, options_.timeout);
   // Fill service-wide initial metadata.
-  for (const auto& header_value : parent_.initial_metadata_) {
-    headers_message_->headers().addCopy(Http::LowerCaseString(header_value.key()),
-                                        header_value.value());
+  if (options_.parent_context.stream_info != nullptr) {
+    ASSERT(metadata_parser_ != nullptr);
+    metadata_parser_->evaluateHeaders(headers_message_->headers(),
+                                      *options_.parent_context.stream_info);
+  } else {
+    for (const auto& header_value : parent_.initial_metadata_) {
+      headers_message_->headers().addCopy(Http::LowerCaseString(header_value.key()),
+                                          header_value.value());
+    }
```

Comment:
Maybe better to collapse these two together, it's a bunch of extra complexity having to follow the two branches.